src/gw_gpu_kernel.cpp

```
@@ -271,9 +313,17 @@ namespace green::gpu {
         ss << "can create: " << _nqkpt << " qkpts in parallel" << std::endl;
       }
       std::cout << ss.str();
-      if (_nqkpt == 0) throw std::runtime_error("not enough memory to create qkpt. Please reduce nt_batch");
-      if (_nqkpt == 1 && _ink != 1 && !utils::context.global_rank)
-        std::cerr << "WARNING: ONLY ONE QKPT CREATED. LIKELY CODE WILL BE SLOW. REDUCE NT_BATCH" << std::endl;
+      // NOTE: implement checks again in case nt_batch is specified by user
+      if (_nqkpt == 0 && _nt_batch == 1)
+        throw std::runtime_error("Not enough memory to create qkpt even with nt_batch = 1. Cannot run application on GPU.");
+      if (_nqkpt == 0)
+        throw std::runtime_error("Not enough memory to create qkpt. Please reduce nt_batch");
```

If nt_batch is too large, should we raise an error and exit, or automatically fall back to optimized values?”

---

Good point! My logic is that if the user is specifying `nt_batch` manually, they want to override optimization, say, for debugging. Eventually, I would want no user intervention with `nt_batch`. What do you think? Could use input from @egull as well.

---

I wouldn't fall back to the optimized values, since likely the user knew what they were doing.